Start of 'flow' app

d3zd3z opened the pull request:
This "flow" app is intended to demonstrate the entire flow of cryptographic information through this process.  It is implemented in Python, using the basic cryptography library.

This first implementation contains a command to generate the initial CA. The Python cryptography library has the advantage of making everything explicit.

Signed-off-by: David Brown <david.brown@linaro.org>

microbuilder commented on tools/flow/keys.py:
```diff
@@ -130,7 +130,8 @@ def gen(self, name, ca):
             x509.NameAttribute(NameOID.ORGANIZATION_NAME, u'Linaro Ltd'),
-            x509.NameAttribute(NameOID.ORGANIZATIONAL_UNIT_NAME, u'LinaroCA Device Cert - Signing'),
+            x509.NameAttribute(NameOID.ORGANIZATIONAL_UNIT_NAME,
+                               u'LinaroCA Device Cert - ' + purpose),
```
Didn’t we want to get away from ‘LinaroCA’?

d3zd3z replied:
Yeah, this should probably be something that indicates its "demo" nature.

Also, what was the reason we put the "Signing" part in the OU, but not in the purpose bits of the cert? Or have we just not gotten around to that?

microbuilder replied:
That sounds like a mistake or oversight.

d3zd3z committed:
flow: Create a flow example for COSE_Encrypt decryption

Create an example workflow demonstrating how to decrypt the example COSE_Encrypt
packets from the cose-examples repo.  This includes a test which can be
exercised by running `cargo test` in the `tools/flow` directory.  This should be
fairly close to what is needed for decryption.  There are some missing error
checks that will result in panics instead of errors being returned.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Add rand_core dependency

This dependency allows us to generate our own keys for tests.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Some missing files

These files were missing from the flow example, and are contain most of the
functionality.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Implement cose_encrypt

Implement COSE_encrypt support, ensuring that the result can be decoded by the
decrypt function. The resulting packet has a minimum of information in it, and
will need to gain support for key id's and possibly session identifier
information.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Add workflow to cargo test the flow app

Add a github workflow to run the flow code through cargo test.  This should
catch compilation errors and make sure the tests pass.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Add verification support for COSE_Sign1

Decode the sign1 signature example packet, and verify that the signature
matches.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Make 'Key' into an enum

Currently, we always know the secret key when handling keys. However, there will
be cases, where we only know the public key.  Change this Key type into an enum
to allow for this expansion.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Add signing support

Implement and test COSE_Sign1 support with ECDSA signatures.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Make recipient context optional

For the full COSE_Encrypt packet, the context was decoded as hex, because this
was useful for ensuring we got the context correct. However, this doesn't work
as the context is optional, and some packets, such as COSE_Encrypt0, do not have
a context.  The proper fix for this is to make it an `Option<Vec<u8>>` and write
an encoding module, like 'hex' that handles the field being optional.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Implement decryption of COSE_Encrypt0

Decrypt the COSE_Encrypt0 packets, and implement reading the cose-example data
for this.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Implement and test COSE_Encrypt0 generation

Test generating COSE_Encrypt0 packets and verifying that we can decrypt and
verify them.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Clean up warnings when building main

There are numerous warnings about unused symbols that come when building main.
Until these are used by main, suppress these warnings.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: rmstfmt main

Run rustfmt to standardize the formatting in this file.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Wrap Key with a struct for additional information

We will want to store identities, such as a key-id with each key. Instead of
putting this in each enum field, make the outer type a struct, with the enum as
a single field within it.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Add key_id to the Key structure

Create a field to hold the key_id in the Key structure.  When we are getting the
keys from an example, populate this with the `kid` field in the example key.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Read x509 certificates into a public key

Parse an x509 certificate, returning a Key that has the public key, and a key-id
derived from the certificate.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Allow a private key to be used as well

If we detect a .pk8 file next to a certificate file, load that key as the
private key for this certificate.  We don't currently check that the private and
public keys go together, so this could potentially result in confusing errors if
the wrong private key is in the private key file.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Ensure that the private key matches the certificate

Compare the public key generated from the private key with the one in the
certificate, and return an error if this private key doesn't go with this public
key.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Allow access to the secret key bytes

In order to be able to share these keys, allow access to the raw bytes of the
secret key.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Start of 'start-session' command

This command will start a new crypto session in the flow.  There are some pieces
missing (such as key-ids and some protected fields to identify the session), and
the packet is currently just printed out.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Auto updated Cargo.lock file

After adding pem support, there are additional dependencies, and a newer version
of one package.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Write session and state outfile

Write the session and the state file out.  Currently, the state file is defined
in cbor, although we may want something more resilient to structure change.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Add encrypt command

The encrypt command encrypts a block of data using a given session key.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Implement payload decryption

Implements the checking and decryption of the payload from the session file.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Make the need for a private key explicit

Instead of always just looking for a parallel .pk8 file, and loading it if
present, require the caller of `Key::from_cert_file` to explicitly state whether
the private key will be needed. It will return an error if the private key is
not present and is needed. This will help clarify, in the code, which cases of a
key just need the public key from the certificate and which times need the
private key.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Wrap CoseError to handle errors better

The CoseError error return type doesn't implement std::error::Error, which means
it can't be used with the `?` construct in Rust.  Write a small wrapper for this
to allow it to be used in these cases, and handle the resulting errors.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Handle some errors better

In the cases for errors that do implement std::error::Error, use `?` to return
the error instead of just panicing. There are still numerous error types in the
cryptography package that do not implement Error, so there are some failures
that will still panic.

Signed-off-by: David Brown <david.brown@linaro.org>

microbuilder commented on tools/flow/README.md:
```diff
@@ -0,0 +1,21 @@
+# Setting things up.
+
+Flow doesn't try to recreate the key creation parts of what Linaro CA
```
LITE Bootstrap Server now (https://github.com/Linaro/lite_bootstrap_server)

d3zd3z replied:
Fixed this.

d3zd3z committed:
flow: Add a tagging module to help with COSE CBOR tags

The cose and ciborium libraries don't make it all that easy to add CBOR tags to
the COSE packets, requiring the packet to be returned as a Value, which is
wasteful.  Create a small module that is able to tag cbor packets, by calling
`encode` before writing data from a packet, and `decode` to extract the tag and
remaining packets.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Use the tagging module in the unit tests

Instead of manually decoding the cbor tags in the cose-examples tests, use the
new tagging library.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: rustfmt the keys file

This file has drifted from rustfmt formatting. Run the tool to clean up the
formatting.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Add tags to the top-level COSE packets

Add a tag to the top-level COSE packets so we can distinguish between the Sign1
and the Encrypt0 of the payload.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Generate and store a session ID

The session id is a sequence of bytes that uniquely identifies a given session.
We will use our own tag for this for now, since none of the existing ones seem
appropriate.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Move from anyhow to thiserror

Use our own error type instead of lots of ad-hoc errors.  This still would need
a lot of cleanup before this crate could really be used, and unless
https://github.com/google/coset/issues/70 gets fixed, we still need to manually
wrap those errors in order to fix the issue (it might be possible to just
implement from for it).

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Allow CoseError to be a regular error

It turns out that the `std` feature on coset is not the default, and enabling
this allows the `CoseError` to be used as a regular Error type in Rust.  Enable
this feature, and remove the wrapper workaround we had in place because of this.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Clean up external errors

Clean up the various wrapped errors.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: rustfmt on the keys file

File contents had gotten out of date from canonical.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Put real key-id into COSE_Sign1 packet

Extract the key id from the certificate, and use as the keyid in the COSE
packet.  To do this, we need to pass the Key in itself, rather than just the
public key taken from it.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Add meaningful content type to sign message

Allow the caller to set what the content type of the payload is, we can use this
to ensure that we are reading the right type of packet.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Add simple encrypt/decrypt commands

These commands will simply encrypt a piece of data with the signed/encrypted
payload.  The application type is different to better distinguish these.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Use proper key-id for key idenfitication

Be sure the use the right key-id for the COSE_Sign1 packet.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Update readme with examples of usage

Populate the readme with an annotated description of how the tool can be used.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Make the filenames in the README more consistent

Use the same names in the first section as in the second section to make it
easier to follow.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Further README clarifications

More correlation between filenames in the first section with the later example.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
flow: Remove non-functional commands

The flow utility has two commands `test` and `gen` that don't have meaningful
implementations, and more represent false starts in the implementation. Remove
these to avoid confusion with the usage of the program.

Signed-off-by: David Brown <david.brown@linaro.org>
